model/inline/src/main/scala/aqua/model/inline/tag/StreamRestrictions.scala

```diff
@@ -0,0 +1,41 @@
+package aqua.model.inline.tag
+
+import aqua.model.inline.state.{Arrows, Config, Exports, Mangler}
+import aqua.model.{OpModel, RestrictionModel}
+import aqua.types.StreamType
+
+import cats.data.{Chain, State}
+
+object StreamRestrictions {
+
+  // restrict streams that are generated in a tree
+  def restrictStreams[S: Mangler: Exports: Arrows: Config](
+    childrenToModel: Chain[OpModel.Tree] => OpModel.Tree
+  )(children: State[S, Chain[OpModel.Tree]]): State[S, OpModel.Tree] =
+    restrictStreamsAround(true)(children.map(childrenToModel))
+
+  /**
+   * Restrict streams that are generated in a tree
+   * @param wrapInside if true, restrict around children in a tree, if false - restrict around a model
+   */
+  def restrictStreamsAround[S: Mangler: Exports: Arrows: Config](wrapInside: Boolean = false)(
+    getTree: State[S, OpModel.Tree]
+  ): State[S, OpModel.Tree] = {
+    for {
+      streamsBefore <- Exports[S].streams
+      tree <- getTree
+      streamsAfter <- Exports[S].streams
+      streams = streamsAfter.removedAll(streamsBefore.keySet)
+      _ <- Exports[S].deleteStreams(streams.keySet)
```

Comment:
I would suggest replacing function `deleteStreams` with a function `streamScope[T](inside: State[S, T]): State[S, (T, Streams)]` that does what this five lines do. Scoping is easier to maintain than explicit delete of information.